### PR TITLE
Fix a bug where the chain is forked when a branch point doesn't exist

### LINKED ISSUE #651: NullReferenceException in BlockChain<T>.Swap
```
System.NullReferenceException: Object reference not set to an instance of an object
  at Libplanet.Blockchain.BlockChain`1[T].Swap (Libplanet.Blockchain.BlockChain`1[T] other, System.Boolean render) [0x00231] in <4c7d57c4ad7b4752aa1435fdd58512d0>:0 
  at Libplanet.Net.Swarm`1+<SyncPreviousBlocksAsync>d__103[T].MoveNext () [0x001bf] in <4c7d57c4ad7b4752aa1435fdd58512d0>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at Libplanet.Net.Swarm`1+<AppendBlocksAsync>d__104[T].MoveNext () [0x00103] in <4c7d57c4ad7b4752aa1435fdd58512d0>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at Libplanet.Net.Swarm`1+<ProcessBlockHashes>d__102[T].MoveNext () [0x002f9] in <4c7d57c4ad7b4752aa1435fdd58512d0>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at Libplanet.Net.Swarm`1+<ProcessMessageAsync>d__101[T].MoveNext () [0x0021e] in <4c7d57c4ad7b4752aa1435fdd58512d0>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <1f0c1ef1ad524c38bbc5536809c46b48>:0 
  at Libplanet.Net.Swarm`1+<>c__DisplayClass110_0+<<ReceiveMessage>b__0>d[T].MoveNext () [0x0007b] in <4c7d57c4ad7b4752aa1435fdd58512d0>:0 
```

### LINKED ISSUE #652: Libplanet.Blocks.InvalidBlockIndexException when Tip index is 0 and block index is 1
```
11/05/2019 07:41:45 +00:00: A message has parsed: "Libplanet.Net.Messages.BlockHashes", from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"
11/05/2019 07:41:45 +00:00: Bucket already contains ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"]
11/05/2019 07:41:45 +00:00: Adding ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"] to routing table. (evicted : null)
11/05/2019 07:41:45 +00:00: Trying to GetBlocksAsync() using 1 hashes.
11/05/2019 07:41:45 +00:00: Adding request (fa007429-aaef-4306-8202-c685b7e93f4b) to queue...
11/05/2019 07:41:45 +00:00: Request fa007429-aaef-4306-8202-c685b7e93f4b taken.
11/05/2019 07:41:45 +00:00: Request Added. waiting for reply...
11/05/2019 07:41:45 +00:00: Trying to send "Libplanet.Net.Messages.GetBlocks" to "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:45 +00:00: A message "Libplanet.Net.Messages.GetBlocks" sent.
11/05/2019 07:41:45 +00:00: A raw message (5 frames) has replied.
11/05/2019 07:41:45 +00:00: A reply has parsed: "Libplanet.Net.Messages.Blocks" from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"
11/05/2019 07:41:45 +00:00: Received ["Libplanet.Net.Messages.Blocks"] from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:45 +00:00: Bucket already contains ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"]
11/05/2019 07:41:45 +00:00: Adding ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"] to routing table. (evicted : null)
11/05/2019 07:41:45 +00:00: Received 1 blocks from "911509101D3bd1992376B1e8eE7fB82e60c8EF35".
11/05/2019 07:41:45 +00:00: GetBlocksAsync() complete.
11/05/2019 07:41:45 +00:00: Trying to fill up previous blocks...
11/05/2019 07:41:45 +00:00: Trying to find branchpoint...
11/05/2019 07:41:45 +00:00: Locator's count: 5
11/05/2019 07:41:45 +00:00: Adding request (90ab7815-322f-42e5-93e5-e76db25dd217) to queue...
11/05/2019 07:41:45 +00:00: Request Added. waiting for reply...
11/05/2019 07:41:45 +00:00: Request 90ab7815-322f-42e5-93e5-e76db25dd217 taken.
11/05/2019 07:41:45 +00:00: Trying to send "Libplanet.Net.Messages.GetBlockHashes" to "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:45 +00:00: A message "Libplanet.Net.Messages.GetBlockHashes" sent.
11/05/2019 07:41:45 +00:00: Waiting for new request...
11/05/2019 07:41:45 +00:00: A raw message [frame count: 6] has received.
11/05/2019 07:41:45 +00:00: A message has parsed: "Libplanet.Net.Messages.BlockHashes", from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"
11/05/2019 07:41:45 +00:00: Bucket already contains ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"]
11/05/2019 07:41:45 +00:00: Adding ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"] to routing table. (evicted : null)
11/05/2019 07:41:45 +00:00: Trying to GetBlocksAsync() using 1 hashes.
11/05/2019 07:41:45 +00:00: Adding request (c83ab8ce-46c9-461d-92d0-10a5ede46aa9) to queue...
11/05/2019 07:41:45 +00:00: Request Added. waiting for reply...
11/05/2019 07:41:45 +00:00: Request c83ab8ce-46c9-461d-92d0-10a5ede46aa9 taken.
11/05/2019 07:41:45 +00:00: Trying to send "Libplanet.Net.Messages.GetBlocks" to "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:45 +00:00: A message "Libplanet.Net.Messages.GetBlocks" sent.
11/05/2019 07:41:45 +00:00: A raw message (9 frames) has replied.
11/05/2019 07:41:45 +00:00: A reply has parsed: "Libplanet.Net.Messages.BlockHashes" from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"
11/05/2019 07:41:45 +00:00: Received ["Libplanet.Net.Messages.BlockHashes"] from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:45 +00:00: Bucket already contains ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"]
11/05/2019 07:41:45 +00:00: Adding ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"] to routing table. (evicted : null)
11/05/2019 07:41:45 +00:00: Branchpoint is "8e7f30d61bb62dda61d00f703bb288fd200d5e6a42705edf50ed1dec78e9f6c9".
11/05/2019 07:41:45 +00:00: Forking needed. Trying to fork...
11/05/2019 07:41:45 +00:00: A raw message (5 frames) has replied.
11/05/2019 07:41:45 +00:00: A reply has parsed: "Libplanet.Net.Messages.Blocks" from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"
11/05/2019 07:41:45 +00:00: Received ["Libplanet.Net.Messages.Blocks"] from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:45 +00:00: Bucket already contains ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"]
11/05/2019 07:41:45 +00:00: Adding ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"] to routing table. (evicted : null)
11/05/2019 07:41:45 +00:00: Received 1 blocks from "911509101D3bd1992376B1e8eE7fB82e60c8EF35".
11/05/2019 07:41:45 +00:00: GetBlocksAsync() complete.
11/05/2019 07:41:45 +00:00: Waiting for new request...
11/05/2019 07:41:45 +00:00: Trying to fork chain at "8e7f30d61bb62dda61d00f703bb288fd200d5e6a42705edf50ed1dec78e9f6c9"(prevId: b2ca5123-ca63-44ab-8c7f-a80f3dbd3396) (forkedId: 2d86134a-bd4f-409b-88c9-37e27b9d1d3f)
11/05/2019 07:41:45 +00:00: Waiting for new request...
11/05/2019 07:41:46 +00:00: Forking complete.
11/05/2019 07:41:46 +00:00: Trying to fill up previous blocks...
11/05/2019 07:41:46 +00:00: Required hashes (count: 4). (tip: 8e7f30d61bb62dda61d00f703bb288fd200d5e6a42705edf50ed1dec78e9f6c9)
11/05/2019 07:41:46 +00:00: Adding request (4633ecd1-1a40-43c1-a615-d02147b69e51) to queue...
11/05/2019 07:41:46 +00:00: Request Added. waiting for reply...
11/05/2019 07:41:46 +00:00: Request 4633ecd1-1a40-43c1-a615-d02147b69e51 taken.
11/05/2019 07:41:46 +00:00: Trying to send "Libplanet.Net.Messages.GetBlocks" to "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:46 +00:00: A message "Libplanet.Net.Messages.GetBlocks" sent.
11/05/2019 07:41:46 +00:00: A raw message (8 frames) has replied.
11/05/2019 07:41:46 +00:00: A reply has parsed: "Libplanet.Net.Messages.Blocks" from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"
11/05/2019 07:41:46 +00:00: Received ["Libplanet.Net.Messages.Blocks"] from "0x911509101D3bd1992376B1e8eE7fB82e60c8EF35"...
11/05/2019 07:41:46 +00:00: Bucket already contains ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"]
11/05/2019 07:41:46 +00:00: Adding ["0x911509101D3bd1992376B1e8eE7fB82e60c8EF35.Unspecified/host.docker.internal:59875.1"] to routing table. (evicted : null)
11/05/2019 07:41:46 +00:00: Received 4 blocks from "911509101D3bd1992376B1e8eE7fB82e60c8EF35".
11/05/2019 07:41:46 +00:00: Trying to append block[2462eaa25f767e0cf153f46de7879bf27085b42ad6c1211d2a53d345be060000]...
11/05/2019 07:41:46 +00:00: Trying to append block 1: "2462eaa25f767e0cf153f46de7879bf27085b42ad6c1211d2a53d345be060000"
11/05/2019 07:41:46 +00:00: Append failed. The block is invalid.
Libplanet.Blocks.InvalidBlockIndexException: the expected block index is 6, but its index is 1'
```
It shouldn't throw `InvalidBlockIndexException` because the tip index of the chain is 0 and the block index to append is 1.

### PR DESCRIPTION
**Pull request by earlbread**
This fixes a bug where the chain is forked when a branch point doesn't exist in the chain.

Resolves #651 
Resolves #652

### CODE REVIEW

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 1]
```diff
@@ -2249,6 +2248,76 @@ public async Task HandleReorgInSynchronizing()
...
+                await minerChainB.MineBlock(_fx1.Address1);
+                minerSwarmA.BroadcastBlocks(new[] { b3 });
+                await receiverSwarm.BlockReceived.WaitAsync();
+                Assert.Equal(receiverChain.Tip, minerChainA.Tip);
+            }
+            finally
+            {
+                await minerSwarmA.StopAsync();
+                await minerSwarmB.StopAsync();
+                await receiverSwarm.StopAsync();
```
Comment (longfin) [inline Libplanet.Tests/Net/SwarmTest.cs]: I guess we should `.Dispose()` against created `Swarm<T>`s.

**Comment by earlbread** [comment]
I added the changelog.

File: Libplanet/Blockchain/BlockChain.cs [hunk 2]
```diff
@@ -972,6 +972,22 @@ internal IReadOnlyList<ActionEvaluation> EvaluateActions(Block<T> block)
...
         internal BlockChain<T> Fork(HashDigest<SHA256> point)
         {
+            if (!Contains(point))
+            {
+                throw new ArgumentException(
+                    $"The block [{point}] doesn't exist.",
+                    nameof(point));
+            }
+
+            Block<T> pointBlock = this[point];
```
Comment (limebell) [inline Libplanet/Blockchain/BlockChain.cs]: Shouldn't this reference done after `_rwlock.EnterReadLock`?

Comment (earlbread) [reply Libplanet/Blockchain/BlockChain.cs]: Since `BlockChain<T>[HashDigest<SHA256>]` has its own lock, I moved it here.

**Pull request merged**